### ISSUE TITLE
Integrating datasets with SCTransform in Seurat v5

### ISSUE THREAD

**Issue by Zack-G-123** [satijalab/seurat#7542]
I have previously used Seurat v4 for integrating across samples with SCTransform, and would like to use this method in Seurat v5. However, I was hoping to take advantage of the Assay5 structure, and well as the IntegrateLayers() functionality to test out various integration methods. Looking around at other forum posts, I've seen several mentions of v5 being compatible with SCT, but I can't get a clear idea of the proper way to use it with IntegrateLayers().

My old workflow would look something like this (example with 3 datasets generated from the Read10Xx command):
obj1 <- SCTransform(obj1, vst.flavor = "v2")
obj2 <- SCTransform(obj2, vst.flavor = "v2")
obj3 <- SCTransform(obj3, vst.flavor = "v2")
sample.list <- list(obj1, obj2, obj3)
features <- SelectIntegrationFeatures(object.list = sample.list, nfeatures = 3000)
sample.list <- PrepSCTIntegration(object.list = sample.list, anchor.features = features)
anchors <- FindIntegrationAnchors(object.list = sample.list, normalization.method = "SCT", anchor.features = features)
combined<- IntegrateData(anchorset = anchors, normalization.method = "SCT")
combined<- RunPCA(combined)
combined<- RunUMAP(combined, dims = 1:30)
combined<- FindNeighbors(combined, dims = 1:30)
combined<- FindClusters(combined, resolution = 0.6)

With the new v5 changes, I've been trying various ways to replicate this pipeline. The only one that seems to be working is:
combined.v5 <- merge(obj1, list(obj2, obj3))
combined.v5 <- SCTransform(combined.v5, vst.flavor = "v2")
combined.v5 <- RunPCA(combined.v5)
combined.v5 <- IntegrateLayers(combined.v5, CCAIntegration, normalization.method = "SCT")
combined.v5 <- RunUMAP(combined.v5, reduction = "integrated.dr", dims = 1:30)
combined.v5 <- FindNeighbors(combined.v5, dims = 1:30, reduction = "integrated.dr")
combined.v5 <- FindClusters(combined.v5, resolution = 0.6)

However, there are a couple things that are unclear to me about this.
1) Running SCTransform on an RNA assay with class Assay5 returns an SCT assay of class SCTAssay, which does not appear to contain layers. I'm unclear how this will then interact with the IntegrateLayers() command and if layers need to be established first / the assay needs to be an Assay5 to integrate properly.
2) Similar to the above, if I first convert the SCT assay to an Assay5 class with layers based on original sample, e.g. using: combined[["SCT"]] <- split(combined[["SCT"]], f = combined$sample), RunPCA throws an error if I try to run it on the SCT assay. When (if ever) should an SCT assay be converted from class SCTAssay to Assay5?
3) When running PCA in the new pipeline, wouldn't this be calculating PCs based on non-integrated scaled data in the SCT assay that has only been normalized within-samples? Whereas in the original pipeline it would use data that has already been scaled/corrected across samples?
5) Finally, under this workflow I want to make sure I am proceeding to downstream DEG testing correctly. Should I still run PrepSCTFindMarkers() prior to analyses? In vignettes using standard log normalization with Seurat v5, I also see that JoinLayers() is required prior to DE testing - is this the case with the SCTransform pipeline in v5 as well?

TL;DR, starting from a list of unprocessed Seurat objects, what is the correct workflow for performing SCT normalization in a way that allows a) testing different integration methods with IntegrateLayers() but using SCTransformed data, and b) proceeding to downstream differential expression testing using SCT-corrected counts

**Comment by Lucyyang1991** [comment]
I have similar confusion.
And about the workflow of integrating datasets with SCTransfrom v2, I first read the issue #7342 , I found this not workout.
````
obj <- LoadData("ifnb")
obj[["RNA"]] <- split(obj[["RNA"]], f = obj$stim)
obj = PercentageFeatureSet(object = obj, col.name = "percent.mt") %>%
  SCTransform(object = ., vars.to.regress = "percent.mt", vst.flavor = "v2", method = "glmGamPoi") %>% 
  RunPCA()
obj <- IntegrateLayers(object = obj, method = HarmonyIntegration, 
  orig.reduction = "pca", new.reduction = 'harmony', 
  assay = "SCT", verbose = FALSE)
````
Then, I changed into
```
obj <- IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony", normalization.method = "SCT")
```
things get work out.
But, when I follow the manual of  `scVIIntegration {SeuratWrappers}`,
```
obj <- IntegrateLayers(
  object = obj, method = scVIIntegration, orig.reduction = "pca",
  new.reduction = "integrated.scvi", assay = "SCT",
  conda_env = "C:/ProgramData/miniforge3/")
```
I got the error below:
```
Error in UseMethod(generic = "JoinLayers", object = object) : 
  no applicable method for 'JoinLayers' applied to an object of class "c('SCTAssay', 'Assay', 'KeyMixin')"
```
Then, I checked
```
> scVIIntegration
function (object, groups = NULL, features = NULL, layers = "counts", 
    conda_env = NULL, new.reduction = "integrated.dr", ndims = 30, 
    nlayers = 2, gene_likelihood = "nb", max_epochs = NULL, ...) 
{
    reticulate::use_condaenv(conda_env, required = TRUE)
    sc <- reticulate::import("scanpy", convert = FALSE)
    scvi <- reticulate::import("scvi", convert = FALSE)
    anndata <- reticulate::import("anndata", convert = FALSE)
    scipy <- reticulate::import("scipy", convert = FALSE)
    object <- JoinLayers(object = object, layers = "counts")
    adata <- sc$AnnData(X = scipy$sparse$csr_matrix(Matrix::t(LayerData(object, 
        layers = "counts")[features, ])), obs = groups, var = object[][features, 
        ])
    scvi$model$SCVI$setup_anndata(adata, batch_key = "group")
    model = scvi$model$SCVI(adata = adata, n_latent = as.integer(x = ndims), 
        n_layers = as.integer(x = nlayers), gene_likelihood = gene_likelihood)
    if (is.null(max_epochs)) {
        max_epochs <- reticulate::r_to_py(x = max_epochs)
    }
    else {
        max_epochs <- as.integer(x = max_epochs)
    }
    model$train(max_epochs = max_epochs)
    latent = model$get_latent_representation()
    latent <- as.matrix(latent)
    rownames(latent) <- reticulate::py_to_r(adata$obs$index$values)
    colnames(latent) <- paste0(new.reduction, "_", 1:ncol(latent))
    suppressWarnings(latent.dr <- CreateDimReducObject(embeddings = latent, 
        key = new.reduction))
    output.list <- list(latent.dr)
    names(output.list) <- new.reduction
    return(output.list)
}
<bytecode: 0x0000020417aa7088>
<environment: namespace:SeuratWrappers>
attr(,"Seurat.method")
[1] "integration"
```
It seems that parameter of `assay = "SCT"` was not included in `scVIIntegration`?

And here is my sessioninfo:
```
R version 4.3.1 (2023-06-16 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=Chinese (Simplified)_China.utf8  LC_CTYPE=Chinese (Simplified)_China.utf8    LC_MONETARY=Chinese (Simplified)_China.utf8
[4] LC_NUMERIC=C                                LC_TIME=Chinese (Simplified)_China.utf8    

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] dplyr_1.1.2              future_1.33.0            patchwork_1.1.2          ggplot2_3.4.2            Azimuth_0.4.6.9004      
 [6] shinyBS_0.61.1           SeuratWrappers_0.3.19    pbmcsca.SeuratData_3.0.0 pbmcref.SeuratData_1.0.0 SeuratData_0.2.2.9001   
[11] Seurat_4.9.9.9050        SeuratObject_4.9.9.9086  sp_2.0-0                

loaded via a namespace (and not attached):
  [1] fs_1.6.2                          ProtGenerics_1.32.0               matrixStats_1.0.0                 spatstat.sparse_3.0-2            
  [5] bitops_1.0-7                      DirichletMultinomial_1.42.0       TFBSTools_1.38.0                  httr_1.4.6                       
  [9] RColorBrewer_1.1-3                ggsci_3.0.0                       tools_4.3.1                       sctransform_0.3.5                
 [13] utf8_1.2.3                        R6_2.5.1                          DT_0.28                           lazyeval_0.2.2                   
 [17] uwot_0.1.16                       rhdf5filters_1.12.1               withr_2.5.0                       prettyunits_1.1.1                
 [21] gridExtra_2.3                     progressr_0.13.0                  cli_3.6.1                         Biobase_2.60.0                   
 [25] spatstat.explore_3.2-1            fastDummies_1.7.3                 EnsDb.Hsapiens.v86_2.99.0         shinyjs_2.1.0                    
 [29] labeling_0.4.2                    spatstat.data_3.0-1               readr_2.1.4                       ggridges_0.5.4                   
 [33] pbapply_1.7-2                     Rsamtools_2.16.0                  R.utils_2.12.2                    harmony_0.1.1                    
 [37] parallelly_1.36.0                 BSgenome_1.68.0                   rstudioapi_0.15.0                 RSQLite_2.3.1                    
 [41] generics_0.1.3                    BiocIO_1.10.0                     gtools_3.9.4                      ica_1.0-3                        
 [45] spatstat.random_3.1-5             googlesheets4_1.1.1               GO.db_3.17.0                      Matrix_1.6-0                     
 [49] fansi_1.0.4                       S4Vectors_0.38.1                  abind_1.4-5                       R.methodsS3_1.8.2                
 [53] lifecycle_1.0.3                   yaml_2.3.7                        SummarizedExperiment_1.30.2       glmGamPoi_1.12.2                 
 [57] rhdf5_2.44.0                      BiocFileCache_2.8.0               Rtsne_0.16                        grid_4.3.1                       
 [61] blob_1.2.4                        ifnb.SeuratData_3.1.0             promises_1.2.0.1                  shinydashboard_0.7.2             
 [65] crayon_1.5.2                      miniUI_0.1.1.1                    lattice_0.21-8                    cowplot_1.1.1                    
 [69] annotate_1.78.0                   GenomicFeatures_1.52.1            KEGGREST_1.40.0                   knitr_1.43                       
 [73] pillar_1.9.0                      GenomicRanges_1.52.0              rjson_0.2.21                      future.apply_1.11.0              
 [77] codetools_0.2-19                  fastmatch_1.1-3                   leiden_0.4.3                      glue_1.6.2                       
 [81] data.table_1.14.8                 remotes_2.4.2                     vctrs_0.6.3                       png_0.1-8                        
 [85] spam_2.9-1                        cellranger_1.1.0                  gtable_0.3.3                      poweRlaw_0.70.6                  
 [89] cachem_1.0.8                      xfun_0.39                         Signac_1.10.0                     S4Arrays_1.0.4                   
 [93] mime_0.12                         pracma_2.4.2                      survival_3.5-5                    gargle_1.5.1                     
 [97] RcppRoll_0.3.0                    ellipsis_0.3.2                    fitdistrplus_1.1-11               ROCR_1.0-11                      
[101] nlme_3.1-162                      bit64_4.0.5                       progress_1.2.2                    filelock_1.0.2                   
[105] RcppAnnoy_0.0.21                  rprojroot_2.0.3                   GenomeInfoDb_1.36.1               irlba_2.3.5.1                    
[109] KernSmooth_2.23-21                SeuratDisk_0.0.0.9020             colorspace_2.1-0                  seqLogo_1.66.0                   
[113] BiocGenerics_0.46.0               DBI_1.1.3                         tidyselect_1.2.0                  bit_4.0.5                        
[117] compiler_4.3.1                    curl_5.0.1                        hdf5r_1.3.8                       xml2_1.3.5                       
[121] DelayedArray_0.26.6               plotly_4.10.2                     rtracklayer_1.60.0                scales_1.2.1                     
[125] caTools_1.18.2                    lmtest_0.9-40                     rappdirs_0.3.3                    stringr_1.5.0                    
[129] digest_0.6.33                     goftest_1.2-3                     presto_1.0.0                      spatstat.utils_3.0-3             
[133] rmarkdown_2.23                    XVector_0.40.0                    htmltools_0.5.5                   pkgconfig_2.0.3                  
[137] MatrixGenerics_1.12.2             dbplyr_2.3.3                      fastmap_1.1.1                     ensembldb_2.24.0                 
[141] rlang_1.1.1                       htmlwidgets_1.6.2                 shiny_1.7.4.1                     farver_2.1.1                     
[145] zoo_1.8-12                        jsonlite_1.8.7                    BiocParallel_1.34.2               R.oo_1.25.0                      
[149] RCurl_1.98-1.12                   magrittr_2.0.3                    GenomeInfoDbData_1.2.10           dotCall64_1.0-2                  
[153] Rhdf5lib_1.22.0                   munsell_0.5.0                     Rcpp_1.0.11                       reticulate_1.30                  
[157] stringi_1.7.12                    zlibbioc_1.46.0                   MASS_7.3-60                       plyr_1.8.8                       
[161] parallel_4.3.1                    listenv_0.9.0                     ggrepel_0.9.3                     deldir_1.0-9                     
[165] CNEr_1.36.0                       Biostrings_2.68.1                 splines_4.3.1                     tensor_1.5                       
[169] hms_1.1.3                         BSgenome.Hsapiens.UCSC.hg38_1.4.5 igraph_1.5.0                      spatstat.geom_3.2-2              
[173] RcppHNSW_0.4.1                    reshape2_1.4.4                    biomaRt_2.56.1                    stats4_4.3.1                     
[177] TFMPvalue_0.0.9                   XML_3.99-0.14                     evaluate_0.21                     BiocManager_1.30.21              
[181] JASPAR2020_0.99.10                tzdb_0.4.0                        httpuv_1.6.11                     RANN_2.6.1                       
[185] tidyr_1.3.0                       purrr_1.0.1                       polyclip_1.10-4                   scattermore_1.2                  
[189] rsvd_1.0.5                        xtable_1.8-4                      restfulr_0.0.15                   AnnotationFilter_1.24.0          
[193] RSpectra_0.16-1                   later_1.3.1                       googledrive_2.1.1                 viridisLite_0.4.2                
[197] tibble_3.2.1                      memoise_2.0.1                     AnnotationDbi_1.62.2              GenomicAlignments_1.36.0         
[201] IRanges_2.34.1                    cluster_2.1.4                     globals_0.16.2                    here_1.0.1                       
```
Looking forward to the solution！

**Comment by saketkc** [comment]
Hi @Lucyyang1991 , scVIintegration is currently not supported for SCT as input to SCVI are counts - we can think of supporting SCT using corrected counts, but for now please use other integration approaches (besides sCVI) when working with SCT assay.

Here's an example of using SCT for integration using the panc8 dataset available from SeuratData:

```r
library(Seurat) # make sure you are running SeuratV5
options(Seurat.object.assay.version = 'v5')
library(SeuratData)

obj <- LoadData("pbmcsca")
obj <- UpdateSeuratObject(obj)
obj <- subset(obj, nFeature_RNA > 1000)
obj[["RNA"]] <- split(obj[["RNA"]], f = obj$Method)

#obj <- subset(obj, Method == "Smart-seq2", invert=T)
#obj[["RNA"]] <- split(obj[["RNA"]], f = obj$tech)

# run sctransform
obj <- SCTransform(obj, vst.flavor = "v2")
obj <- RunPCA(obj, npcs = 30, verbose = FALSE)

# one-liner to run Integration
obj <- IntegrateLayers(object = obj, method = HarmonyIntegration,
                       orig.reduction = "pca", new.reduction = 'harmony',
                       assay = "SCT", verbose = FALSE)
obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:30)
obj <- FindClusters(obj, resolution = 2, cluster.name = "harmony_clusters")

#
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")
p1 <- DimPlot(
  obj,
  reduction = "umap.harmony",
  group.by = c("Method", "CellType"),
  combine = FALSE
)
patchwork::wrap_plots(p1)
```

**Comment by saketkc** [comment]
Hi @Zack-G-123,

> Running SCTransform on an RNA assay with class Assay5 returns an SCT assay of class SCTAssay, which does not appear to contain layers. I'm unclear how this will then interact with the IntegrateLayers() command and if layers need to be established first / the assay needs to be an Assay5 to integrate properly.

Your observation is correct. SCTAssay is still a `V3` assay. Internally when you pass `assay="SCT"` to `IntegrateLayers` it uses `FetchResiduals` to fetch the residuals for each of the layer in the `counts` slot using the corresponding SCT model. 
That is, when you run `SCTransform` in V5, it runs `sctransform` on each layer separately and stores the model within the `SCTAssay`. When `IntegrateLayers` is called, it calculates residuals on the fly using the stored residuals. 

> Similar to the above, if I first convert the SCT assay to an Assay5 class with layers based on original sample, e.g. using: combined[["SCT"]] <- split(combined[["SCT"]], f = combined$sample), RunPCA throws an error if I try to run it on the SCT assay. When (if ever) should an SCT assay be converted from class SCTAssay to Assay5?

No, please follow the code example I pasted above.

> When running PCA in the new pipeline, wouldn't this be calculating PCs based on non-integrated scaled data in the SCT assay that has only been normalized within-samples? Whereas in the original pipeline it would use data that has already been scaled/corrected across samples?

My response to your first question should answer this. But in this case, PCA is calculated on the residuals across samples. In the original pipeline, if you were to "merge" the two SCT normalized objects, it would do the same (only that you would have to define the variable features yourself). The way this currently works is by using a variant of `SelectIntegrationFeatures` to rank which features are variable 

> Finally, under this workflow I want to make sure I am proceeding to downstream DEG testing correctly. Should I still run PrepSCTFindMarkers() prior to analyses? In vignettes using standard log normalization with Seurat v5, I also see that JoinLayers() is required prior to DE testing - is this the case with the SCTransform pipeline in v5 as well?

Yes, you still need to run `PrepSCTFindMarkers`. In the coming weeks we will release updates to make this faster, so please be on the lookout.

Hope this helps! Please feel free to reopen if you have more questions.

**Comment by Junedays** [comment]
> Hi @Lucyyang1991 , scVIintegration is currently not supported for SCT as input to SCVI are counts - we can think of supporting SCT using corrected counts, but for now please use other integration approaches (besides sCVI) when working with SCT assay.
> 
> Here's an example of using SCT for integration using the panc8 dataset available from SeuratData:
> 
> ```r
> library(Seurat) # make sure you are running SeuratV5
> options(Seurat.object.assay.version = 'v5')
> library(SeuratData)
> 
> obj <- LoadData("pbmcsca")
> obj <- UpdateSeuratObject(obj)
> obj <- subset(obj, nFeature_RNA > 1000)
> obj[["RNA"]] <- split(obj[["RNA"]], f = obj$Method)
> 
> #obj <- subset(obj, Method == "Smart-seq2", invert=T)
> #obj[["RNA"]] <- split(obj[["RNA"]], f = obj$tech)
> 
> # run sctransform
> obj <- SCTransform(obj, vst.flavor = "v2")
> obj <- RunPCA(obj, npcs = 30, verbose = FALSE)
> 
> # one-liner to run Integration
> obj <- IntegrateLayers(object = obj, method = HarmonyIntegration,
>                        orig.reduction = "pca", new.reduction = 'harmony',
>                        assay = "SCT", verbose = FALSE)
> obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:30)
> obj <- FindClusters(obj, resolution = 2, cluster.name = "harmony_clusters")
> 
> #
> obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")
> p1 <- DimPlot(
>   obj,
>   reduction = "umap.harmony",
>   group.by = c("Method", "CellType"),
>   combine = FALSE
> )
> patchwork::wrap_plots(p1)
> ```

Someone said that adding the normalization.method="SCT" in the IntegrateLayers could also perform the SCT, is it true? Like:

sce <- IntegrateLayers(sce, method = HarmonyIntegration, 
                       orig.reduction = "pca", new.reduction = "harmony",normalization.method="SCT")

**Comment by Lucyyang1991** [comment]
@Junedays
Yes, in my last test, `normalization.method="SCT"` works and doesn't throw an error, but I don't know why. 
```
obj <- IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony", normalization.method = "SCT")
```
If you can run `assay = "SCT"` without error, please don't use `normalization.method = "SCT"`.
```
obj <- IntegrateLayers(object = obj, method = HarmonyIntegration,
                       orig.reduction = "pca", new.reduction = 'harmony',
                       assay = "SCT", verbose = FALSE)
```

**Comment by vpbio** [comment]
@saketkc 
> My response to your first question should answer this. But in this case, PCA is calculated on the residuals across samples. In the original pipeline, if you were to "merge" the two SCT normalized objects, it would do the same (only that you would have to define the variable features yourself). The way this currently works is by using a variant of `SelectIntegrationFeatures` to rank which features are variable

Hi, I have a followup question on this explaination . when you say by using a variant of 'SelectIntegrationFeatures' are you talking about 'SelectSCTIntegrationFeatures' OR 'SelectIntegrationFeatures5'  ? only former gives me an output and i'm not sure if it is correct or not because it gives different number of variable features (3 times) than v4 workflow involving SelectIntegrationFeatures on a list of objects.

Thanks

**Comment by erzakiev** [comment]
> ```r
> obj <- LoadData("pbmcsca")
> ```

Error: Cannot find dataset pbmcsca

**Comment by gevro** [comment]
Hello, It still isn't clear to me - when running IntegrateLayers with harmony on data that underwent SCtransform, should IntegrateLayers be run with normalization.method = "SCT" or with assay = "SCT"?

I don't see how normalization.method = "SCT" would work, despite people above saying it does, because normalization.method = "SCT" is not a parameter in the HarmonyIntegration function.
